tests/chainer_tests/test_optimizer.py

```diff
@@ -499,6 +499,122 @@ def test_update_chainerx_gpu(self):
         self.check_update()
 
 
+@testing.backend.inject_backend_tests(
+    None,
+    [
+        # NumPy
+        {},
+        # CuPy
+        {'use_cuda': True, 'cuda_device': 0},
+        {'use_cuda': True, 'cuda_device': 1},
+        # ChainerX
+        {'use_chainerx': True, 'chainerx_device': 'native:0'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:0'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:1'},
+    ])
+@testing.parameterize(*testing.product({
+    'override_pattern': [
+        'generic',  # only update_core() is overridden
+        'cpu_gpu',  # update_core_{cpu,gpu} are overridden
+        'cpu_gpu_chx',  # update_core_{cpu,gpu,chainerx} are overridden
+    ],
+}))
+class TestGradientMethodUpdate(unittest.TestCase):
+    """Ensures UpdateRule's appropriate methods are called, for various
+    override patterns and parameters with various conditions."""
+
+    def create(self, device):
+
+        class MyLink(chainer.Link):
+            def __init__(self):
+                super(MyLink, self).__init__()
+                with self.init_scope():
+                    self.p1 = chainer.Parameter()  # uninitialized
+
+                    self.p2 = chainer.Parameter(  # initialized, with grad
+                        np.array([3, 2], np.float32))
+                    self.p2.grad = np.array([13, 12], np.float32)
+
+                    self.p3 = chainer.Parameter(  # initialized, without grad
+                        np.array([5, 7], np.float32))
+
+        record = []
```

Comment:
Maybe another name like `invocations` or `invocation_records` seems to be fine.

Reply:
How about `call_record`? "call" is more common term used in other tests.

Reply:
Fixed.